mathics/builtin/list/eol.py

```diff
@@ -580,39 +581,74 @@ class First(Builtin):
     <dl>
       <dt>'First[$expr$]'
       <dd>returns the first element in $expr$.
+
+      <dt>'First[$expr$, $def$]'
+      <dd>returns the first element in $expr$ if it exists or $def$ otherwise.
     </dl>
 
     'First[$expr$]' is equivalent to '$expr$[[1]]'.
 
     >> First[{a, b, c}]
      = a
+
+    The first argument need not be a list:
     >> First[a + b + c]
      = a
+
+    However, the first argument must be Nonatomic when there is a single argument:
     >> First[x]
-     : Nonatomic expression expected.
+     : Nonatomic expression expected at position 1 in First[x].
      = First[x]
+
+    Or if it is not, but a second default argument is provided, that is \
+    evaluated and returned:
+
+    >> First[10, 1+2]
+     = 3
+
     >> First[{}]
      : {} has zero length and no first element.
      = First[{}]
+
+    As before, the first argument is empty, but a default argument is given, \
+    evaluate and return the second argument:
+    >> First[{}, 1+2]
+     = 3
     """
 
+    attributes = A_HOLD_REST | A_PROTECTED
     messages = {
-        "normal": "Nonatomic expression expected.",
+        "argt": "First called with `1` arguments; 1 or 2 arguments are expected.",
+        "normal": "Nonatomic expression expected at position 1 in `1`.",
         "nofirst": "`1` has zero length and no first element.",
     }
     summary_text = "first element of a list or expression"
 
-    def eval(self, expr, evaluation):
-        "First[expr_]"
+    # FIXME: the code and the code for Last are similar and can be DRY'd
+    def eval(self, expr, evaluation: Evaluation, expression: Expression):
+        "First[expr__]"
 
         if isinstance(expr, Atom):
-            evaluation.message("First", "normal")
+            evaluation.message("First", "normal", expression)
             return
-        if len(expr.elements) == 0:
+        expr_len = len(expr.elements)
+        if expr_len == 0:
             evaluation.message("First", "nofirst", expr)
             return
+        if expr_len > 2 and expr.head is SymbolSequence:
+            evaluation.message("First", "argt", expr_len)
+            return
+
+        first_elem = expr.elements[0]
+
+        if expr.head == SymbolSequence or (
+            not isinstance(expr, ListExpression)
+            and len == 2
+            and isinstance(first_elem, Atom)
+        ):
+            return expr.elements[1].evaluate(evaluation)
```

Comment:
Just a comment: to call the `evaluate(evaluation)` method at the end is not needed: it is going to be called in the next `rewrite_apply_eval` iteration

Reply:
Ok - thanks. Will remove the explicit call.